File: Problems/Graphs/Clone-Graph/Clone-Graph.cs
using System;
using System.Collections.Generic;

// Using depth-first search - Time: O(n)

public class Node
{
    public int val;
    public IList<Node> neighbors;

    public Node()
    {
        val = 0;
        neighbors = new List<Node>();
    }

    public Node(int _val)
    {
        val = _val;
        neighbors = new List<Node>();
    }

    public Node(int _val, List<Node> _neighbors)
    {
        val = _val;
        neighbors = _neighbors;
    }
}

public class Solution
{
    private Dictionary<Node, Node> m = new Dictionary<Node, Node>();

    public Node CloneGraph(Node node)
    {
        if (node == null)
            return null;

        if (m.ContainsKey(node))
            return m[node];

        Node cpy = new Node(node.val);
        m[node] = cpy;

        foreach (var neighbor in node.neighbors)
        {
            cpy.neighbors.Add(CloneGraph(neighbor));
        }

        return cpy;
    }
}

class Program
{
    static void Main()
    {
        // Constructing the original graph
        var s = new Solution();
        var n1 = new Node(1);
        var n2 = new Node(2);
        var n3 = new Node(3);
        var n4 = new Node(4);

        n1.neighbors.Add(n2);
        n1.neighbors.Add(n4);
        n2.neighbors.Add(n1);
        n2.neighbors.Add(n3);
        n3.neighbors.Add(n2);
        n3.neighbors.Add(n4);
        n4.neighbors.Add(n1);
        n4.neighbors.Add(n3);

        Console.Write("Input: adjList = [");
        foreach (var node in new Node[] { n1, n2, n3, n4 })
        {
            Console.Write("[");
            foreach (var neighbor in node.neighbors)
            {
                Console.Write(neighbor.val + "");
                if (neighbor != node.neighbors[node.neighbors.Count - 1])
                    Console.Write(", ");
            }
            Console.Write("]");
            if (node != n4)
                Console.Write(", ");
        }
        Console.WriteLine("]");

        Solution sol = new Solution();
        var clone = sol.CloneGraph(n1);

        Console.Write("Output: [");
        foreach (var cloneNode in new Node[] {clone, clone.neighbors[0], clone.neighbors[1].neighbors[1], clone.neighbors[1]})
        {
            Console.Write("[");
            foreach (var neighbor in cloneNode.neighbors)
            {
                Console.Write(neighbor.val + "");
                if (neighbor != cloneNode.neighbors[cloneNode.neighbors.Count - 1])
                    Console.Write(", ");
            }
            Console.Write("]");
            if (cloneNode != clone.neighbors[1])
                Console.Write(", ");
        }
        Console.WriteLine("]");
    }
}

File: Problems/Graphs/Clone-Graph/Clone-Graph.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <queue>

// Using depth-first search - Time: O(n)

class Node {
public:
    int val;
    std::vector<Node *> neighbors;

    Node()
    {
        val = 0;
        neighbors = std::vector<Node *>();
    }

    Node(int _val)
    {
        val = _val;
        neighbors = std::vector<Node *>();
    }

    Node(int _val, std::vector<Node *> _neighbors)
    {
        val = _val;
        neighbors = _neighbors;
    }
};

class Solution {
    std::unordered_map<Node *, Node *> m;

public:
    Node *cloneGraph(Node *node)
    {
        if (!node)
            return nullptr;

        if (m.count(node))
            return m[node];

        auto cpy = new Node(node->val);
        m[node] = cpy;

        for (auto &n : node->neighbors)
        {
            cpy->neighbors.push_back(cloneGraph(n));
        }

        return cpy;
    }
};

int main()
{
    // Constructing the original graph
    Node *n1 = new Node(1);
    Node *n2 = new Node(2);
    Node *n3 = new Node(3);
    Node *n4 = new Node(4);

    n1->neighbors = {n2, n4};
    n2->neighbors = {n1, n3};
    n3->neighbors = {n2, n4};
    n4->neighbors = {n1, n3};

    std::cout << "Input: adjList = [";
    for (const auto &node : {n1, n2, n3, n4})
    {
        std::cout << "[";
        for (const auto &neighbor : node->neighbors)
        {
            std::cout << neighbor->val << "";
            if (neighbor != node->neighbors.back())
                std::cout << ", ";
        }
        std::cout << "]";
        if (node != n4)
            std::cout << ", ";
    }
    std::cout << "]" << std::endl;

    Solution sol;
    Node *clone = sol.cloneGraph(n1);

    std::cout << "Output: [";
    for (const auto &cloneNode : {clone, clone->neighbors[0], clone->neighbors[1]->neighbors[1], clone->neighbors[1]})
    {
        std::cout << "[";
        for (const auto &neighbor : cloneNode->neighbors)
        {
            std::cout << neighbor->val << "";
            if (neighbor != cloneNode->neighbors.back())
                std::cout << ", ";
        }
        std::cout << "]";
        if (cloneNode != clone->neighbors[1])
            std::cout << ", ";
    }
    std::cout << "]" << std::endl;

    delete n1;
    delete n2;
    delete n3;
    delete n4;

    return 0;
}
